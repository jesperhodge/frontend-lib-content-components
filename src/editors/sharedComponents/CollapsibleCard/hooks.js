import { useState } from 'react';

/* eslint-disable import/prefer-default-export */
export const useCollapsible = () => {
  const [isCardCollapsibleOpen, setIsCardCollapsibleOpen] = useState(false);
  return {
    isCardCollapsibleOpen,
    toggleCardCollapse: () => setIsCardCollapsibleOpen(!isCardCollapsibleOpen),
  };
};
