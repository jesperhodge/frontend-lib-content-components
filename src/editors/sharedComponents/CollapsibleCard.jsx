import React from 'react';
import { Collapsible, Icon, Card } from '@edx/paragon';
import { KeyboardArrowUp, KeyboardArrowDown } from '@edx/paragon/icons';
import {
  arrayOf, shape, string, node, bool,
} from 'prop-types';
import { showFullCard } from '../containers/ProblemEditor/components/EditProblemView/SettingsWidget/hooks';

const CardSection = ({
  children, none, isCardCollapsibleOpen, summary,
}) => {
  const show = isCardCollapsibleOpen || summary;
  if (!show) { return null; }

  return (
    <Card.Section className="px-4 pb-4 pt-3">
      <Collapsible.Advanced
        open={!isCardCollapsibleOpen}
      >
        <Collapsible.Body className="collapsible-body">
          <span className={`small ${none ? 'text-gray-500' : 'text-primary-500'}`}>{summary}</span>
        </Collapsible.Body>
      </Collapsible.Advanced>
      <Collapsible.Advanced
        open={isCardCollapsibleOpen}
      >
        <Collapsible.Body className="collapsible-body">
          {children}
        </Collapsible.Body>
      </Collapsible.Advanced>
    </Card.Section>
  );
};
CardSection.propTypes = {
  none: bool,
  children: node.isRequired,
  summary: string,
  isCardCollapsibleOpen: bool.isRequired,
};
CardSection.defaultProps = {
  none: false,
  summary: null,
};

export const CollapsibleCard = ({
  title, className, extraSections, children, summary, ...passThroughProps
}) => {
  const { isCardCollapsibleOpen, toggleCardCollapse } = showFullCard();

  return (
    <Card className={`${className} CollapsibleCard border border-light-700 shadow-none`}>
      <Card.Section className="settingsCardTitleSection">
        <Collapsible.Advanced
          open={isCardCollapsibleOpen}
          onToggle={toggleCardCollapse}
        >
          <Collapsible.Trigger className="collapsible-trigger d-flex">
            <span className="flex-grow-1 text-primary-500 x-small">{title}</span>
            <Collapsible.Visible whenClosed>
              <Icon src={KeyboardArrowDown} />
            </Collapsible.Visible>
            <Collapsible.Visible whenOpen>
              <Icon src={KeyboardArrowUp} />
            </Collapsible.Visible>
          </Collapsible.Trigger>
        </Collapsible.Advanced>
      </Card.Section>
      <CardSection {...passThroughProps} isCardCollapsibleOpen={isCardCollapsibleOpen} summary={summary} key={`settingsOption-${title}-children`}>
        {children}
      </CardSection>
      {extraSections.map((section, index) => (
        <>
          {isCardCollapsibleOpen && <hr />}
          {/* eslint-disable-next-line react/no-array-index-key */}
          <CardSection {...passThroughProps} isCardCollapsibleOpen={isCardCollapsibleOpen} key={`CollapsibleCard-${title}-${index}`}>
            {section.children}
          </CardSection>
        </>
      ))}
    </Card>
  );
};
CollapsibleCard.propTypes = {
  title: string.isRequired,
  children: node.isRequired,
  className: string,
  summary: string.isRequired,
  extraSections: arrayOf(shape({
    children: node,
  })),
};
CollapsibleCard.defaultProps = {
  className: '',
  extraSections: [],
};

export default CollapsibleCard;
