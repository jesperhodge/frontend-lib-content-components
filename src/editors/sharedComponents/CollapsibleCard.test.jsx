import React from 'react';
import { shallow } from 'enzyme';
import CollapsibleCard from './CollapsibleCard';

describe('CollapsibleCard', () => {
  describe('default with children', () => {
    const children = (<h1>My test content</h1>);
    test('snapshot: renders correct', () => {
      expect(shallow(<CollapsibleCard title="Settings Option Title" summary="Settings Option Summary">{children}</CollapsibleCard>)).toMatchSnapshot();
    });
  });
  describe('with additional sections', () => {
    const children = (<h1>First Section</h1>);
    const sections = [<h1>Second Section</h1>, <h1>Third Section</h1>];
    test('snapshot: renders correct', () => {
      expect(shallow(
        <CollapsibleCard title="Settings Option Title" summary="Settings Option Summary" extraSections={sections}>
          {children}
        </CollapsibleCard>,
      )).toMatchSnapshot();
    });
  });
});
