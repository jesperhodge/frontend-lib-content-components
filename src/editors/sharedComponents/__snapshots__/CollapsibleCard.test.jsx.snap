// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CollapsibleCard default with children snapshot: renders correct 1`] = `
<Card
  className=" CollapsibleCard border border-light-700 shadow-none"
>
  <Card.Section
    className="settingsCardTitleSection"
  >
    <Advanced
      onToggle={[Function]}
      open={false}
    >
      <Trigger
        className="collapsible-trigger d-flex"
      >
        <span
          className="flex-grow-1 text-primary-500 x-small"
        >
          Settings Option Title
        </span>
        <Visible
          whenClosed={true}
        >
          <Icon />
        </Visible>
        <Visible
          whenOpen={true}
        >
          <Icon />
        </Visible>
      </Trigger>
    </Advanced>
  </Card.Section>
  <CardSection
    isCardCollapsibleOpen={false}
    key="settingsOption-Settings Option Title-children"
    none={false}
    summary="Settings Option Summary"
  >
    <h1>
      My test content
    </h1>
  </CardSection>
</Card>
`;

exports[`CollapsibleCard render snapshot: renders correct 1`] = `
<Card
  className=" CollapsibleCard border border-light-700 shadow-none"
>
  <Card.Section
    className="settingsCardTitleSection"
  >
    <Advanced
      onToggle={[Function]}
      open={false}
    >
      <Trigger
        className="collapsible-trigger d-flex"
      >
        <span
          className="flex-grow-1 text-primary-500 x-small"
        >
          Settings Option Title
        </span>
        <Visible
          whenClosed={true}
        >
          <Icon />
        </Visible>
        <Visible
          whenOpen={true}
        >
          <Icon />
        </Visible>
      </Trigger>
    </Advanced>
  </Card.Section>
  <BodySection
    isCardCollapsibleOpen={false}
    none={false}
    summary="Settings Option Summary"
  >
    <h1>
      My test content
    </h1>
  </BodySection>
</Card>
`;

exports[`CollapsibleCard with additional sections snapshot: renders correct 1`] = `
<Card
  className=" CollapsibleCard border border-light-700 shadow-none"
>
  <Card.Section
    className="settingsCardTitleSection"
  >
    <Advanced
      onToggle={[Function]}
      open={false}
    >
      <Trigger
        className="collapsible-trigger d-flex"
      >
        <span
          className="flex-grow-1 text-primary-500 x-small"
        >
          Settings Option Title
        </span>
        <Visible
          whenClosed={true}
        >
          <Icon />
        </Visible>
        <Visible
          whenOpen={true}
        >
          <Icon />
        </Visible>
      </Trigger>
    </Advanced>
  </Card.Section>
  <CardSection
    isCardCollapsibleOpen={false}
    key="settingsOption-Settings Option Title-children"
    none={false}
    summary="Settings Option Summary"
  >
    <h1>
      First Section
    </h1>
  </CardSection>
  <CardSection
    isCardCollapsibleOpen={false}
    key="CollapsibleCard-Settings Option Title-0"
    none={false}
    summary={null}
  />
  <CardSection
    isCardCollapsibleOpen={false}
    key="CollapsibleCard-Settings Option Title-1"
    none={false}
    summary={null}
  />
</Card>
`;
